### PR TITLE
Add whitelistHosts option.

### PR DESCRIPTION
**Pull request by dracos**
For #1443

### CODE REVIEW

File: packages/server/lib/controllers/proxy.coffee [hunk 1]
```diff
@@ -75,6 +75,20 @@ module.exports = {
 
         return res.status(503).end()
 
+    ## if we have white listed hosts
+    if wlh = config.whitelistHosts
+      ## and url does not match any of our whitelisted hosts
+      if not matched = blacklist.matches(req.proxiedUrl, wlh)
```
Comment (bahmutov) [inline packages/server/lib/controllers/proxy.coffee]: I am confused by the comment and whitelist / blacklist combination here

**Comment by bahmutov** [comment]
@dracos I think the logic between white listed hosts and black listed should be expressed very clearly, also you can run unit tests to see them fail (as they are failing on CI)
```
cd packages/server
npm run test-unit
```

File: packages/server/lib/controllers/proxy.coffee (hunk 1 above)
Comment (dracos) [reply packages/server/lib/controllers/proxy.coffee]: The whitelist is being passed in to this function as `wlh`, this is just a (now badly named) function. `blacklist` (the library) is now more a `*list`-matching library after these changes so presumably should be renamed.

**Comment by jennifer-shehane** [comment]
As gleb mentioned, some of the tests are failing. We also need tests to cover the `whilelistHosts`, specifically in this file: https://github.com/cypress-io/cypress/blob/issue-1350/packages/server/test/unit/config_spec.coffee#L488

Let us know if you have trouble getting the tests running locally.

**Comment by brian-mann** [comment]
Also beyond unit tests, we **must** have e2e tests to cover this config option. You can take the existing blacklist e2e tests and repurpose them.

**Comment by dracos** [comment]
Have renamed library to hostlist, added unit and e2e tests that pass locally. Don't think the failure there now is in anything I've touched.

**Comment by brian-mann** [comment]
Right that failure is not related.

**Comment by itsravenous** [comment]
Any updates on this? A project I'm working on makes calls to a _lot_ of external URLs and our blacklist is very long (and when new endpoints are added we have to remember to add them).

**Comment by flotwig** [comment]
This is something that can be accomplished once #687 is added. Here is one example of how:

```js
function whitelistMatches(url) {
	// ...
}

cy.route('*', (req) => {
	if (!whitelistMatches(req.url)) {
		return req.destroy()
	}

	req.reply()
})
```

`blacklistHosts` will probably be deprecated at some point too, so I'll close this in favor of #687.

**Pull request closed without merging**